packages/toolpad-app/src/toolpad/DomLoader.tsx

```diff
@@ -75,8 +82,43 @@ export type DomAction =
       node: appDom.AppDomNode;
     };
 
+const undoStack: appDom.AppDom[] = [];
+const redoStack: appDom.AppDom[] = [];
+
+const updateUndoStack = throttle((dom: appDom.AppDom) => {
```

Comment:
Maybe we shouldn't have throttling? Don't we have to keep track of every action the user does, even if they do it quickly?

Reply:
I think you answered your own question - once we have atomic updates, we can change implementation. Until then - it's hardly a case that user could manually execute actions faster than 500ms in between them so hopefully it will be fine until we get to atomic updates